# FIX - SEID doesn't match to spec

dos-gruut opened the pull request:
### SEID 명세서 불일치 문제
(기존)
- SEID를 랜덤하게 생성함.
- tx generator를 제작할 당시에 id필드만 있었고, 명세가 없었음

(현재)
- 명세에 따라 GENTSE-N을 갖도록 수정
- getSEID 함수 생성
- gen_tx가 올바른 SEID를 가져오도록 함
- tx.type: digests -> DIGESTS
- tx.rID는 base64로 인코딩

 ### 기타
- tx_send 함수의 매개변수명 변경(p_num -> se_num)
- tx_send 재실행시 매개변수로 se_num을 보내지 않는 버그 수정
- 기본 SE 번호: 0 -> 1
- README 및 howtouse 함수 출력내용 수정

dos-gruut committed:
# FIX - SEID doesn't match to spec

 ### SEID 명세서 불일치 문제
(기존)
- SEID를 랜덤하게 생성함.
- tx generator를 제작할 당시에 id필드만 있었고, 명세가 없었음
(현재)
- 명세에 따라 GENTSE-N을 갖도록 수정
- getSEID 함수 생성
- gen_tx가 올바른 SEID를 가져오도록 함
- tx.type: digests -> DIGESTS
- tx.rID는 base64로 인코딩

 ### 기타
- tx_send 함수의 매개변수명 변경(p_num -> se_num)
- tx_send 재실행시 매개변수로 se_num을 보내지 않는 버그 수정
- 기본 SE 번호: 0 -> 1
- README 및 howtouse 함수 출력내용 수정

morishjs commented on src/tx_generator.js:
```diff
@@ -65,35 +65,38 @@ main();
     ++txid;
 
-    let rID = tools.get64Hash("TX GENERATOR # " + p_num );
+    let rID = tools.getSEID(se_num);
```
rID는 공증서비스를 요청한 사람의 id로 알고 있습니다. 여기서 SEID는 service endpoint의 id입니다.

dos-gruut replied:
rID는 (머저 입장에서) SE 아이디입니다.
명세서에서는 client의 ID를 (SE 입장에서) rID라고 설명하는 부분이 오해의 소지가 있어서 명세서를 수정했습니다.

morishjs commented on src/common.js:
```diff
@@ -155,7 +155,7 @@ const txToBuffer = function(tx){
 	length = pushBufferList(bf_list, length, Buffer.from(tx.txid, 'base64'));
 	length = pushBufferList(bf_list, length, getBufferedTimestamp(tx.time));
-	length = pushBufferList(bf_list, length, Buffer.from(tx.rID, 'base64'));
+	length = pushBufferList(bf_list, length, Buffer.from(tx.rID));
```
156에서는 base64로 인코딩하고, 158줄에는 인코딩안하고 있어서 찾아보니까 tx_generator.js에서 하고 있네요. 

```javascript
length = pushBufferList(bf_list, length, Buffer.from(tx.rID, 'base64'));
```
base64인코딩을 따로하는게 아니라 한쪽에서 하는게 버그 찾기도 수월하고, 코드 읽는데도 좋을 것 같습니다. 이렇게 따로 인코딩을 하면 저처럼 'base64를 어디서 인코딩하지?'하고 코드를 찾아봐야 하거든요. 제 생각엔 이전 코드가 더 좋은 것 같습니다.

dos-gruut replied:
txid =  256bit, base64 string
- 서명용 buffer stream으로 변환하기 위해 base64 -> byte buffer로 변환

rID, cID  = string type, 사람이 식별
- string -> buf로 변환

아이디 지옥...

dos-gruut committed:
# EDIT - 모호한 변수명 수정

 ### 명확한 함수명
 - txToBuffer -> buildSigBuffer
 - RSA 서명을 생성하기 위한 buffer임
 - 이 때, txid는 base64형태임. 서명 생성을 위해 버퍼로 변환

 ### txid 혼동
 - txid는 256bit hex string임
 - tx 증가시키는 정수형 변수명도 txid였음
 - 혼동을 피하고자 정수형 txid -> n_tx 변경

 ### 기타
 - set time 시간 0.1초로 (노드 하나에 10tps)
 - rID / cID 분리 및 cID는 랜덤하게 생성
 - ID는 사람이 식별하는 것
	- rID, cID
 - id는 컴퓨터가 식별하는 것(?)
	- txid

dos-gruut commented:
모호한 표현들이 많아서 이름을 새로 지어주었어요. 다시 검토 부탁합니다!

morishjs commented on src/tx_generator.js:
```diff
@@ -65,41 +65,46 @@ main();
 
     var tx = {};
-	tx.txid = tools.getSHA256(rID + txid);
+	tx.txid = tools.getSHA256(rID + n_tx);
```
indentation

dos-gruut replied:
수정완료

morishjs commented:
@dos-gruut 
>txid = 256bit, base64 string

서명용 buffer stream으로 변환하기 위해 base64 -> byte buffer로 변환
rID, cID = string type, 사람이 식별

string -> buf로 변환

아까 문서 수정하셨다고 하신 부분이 어딘가요? 제가 보는 문서에는 rID가 아직도 base64로 되어있어서요.
https://thevaulters.atlassian.net/wiki/spaces/SGN/pages/49250343/GE5.+Data+Structure

morishjs reviewed (approved):
사소한 수정사항이라서 approve합니다.

dos-gruut commented:
> 아까 문서 수정하셨다고 하신 부분이 어딘가요? 제가 보는 문서에는 rID가 아직도 base64로 되어있어서요.
> https://thevaulters.atlassian.net/wiki/spaces/SGN/pages/49250343/GE5.+Data+Structure

rID = GENTSE-1 = R0VOVFNFLTE=
https://thevaulters.atlassian.net/wiki/spaces/SGN/pages/47186101/GE3.+Message+Modeling#GE3.MessageModeling-MSG_TX=0xB1
자료의 체크섬 (type = "digests") 설명부분
`requestor's  ID` -> `client ID`
 "Data Digest" content 내의 `rID`는 요청한 Client의 것이다.
-> "Data Digest" content 내의 `client ID`는 요청한 Client의 것이다.
